napari/_app_model/context/_layerlist_context.py

```diff
@@ -163,8 +163,11 @@ def _active_is_image_3d(s: LayerSel) -> bool:
     )
 
 
-def _empty_shapes_layer_selected(s: LayerSel) -> bool:
-    return any(x._type_string == 'shapes' and not len(x.data) for x in s)
+def _empty_shapes_layer_selected(s: LayerSel) -> Callable[[], bool]:
+    def fun() -> bool:
```

Comment:
Pinging @Czaki @jni @lucyleeow I had a look in #7055 and we should be able to fix the current test failure (leaking viewer), by replacing this local def with a proper definition and a partial.

While looking at this I realized for the first time that this context key is actually a hybrid of the original context keys that are updated on events, and the new functional context keys that are evaluated on access. To clarify, the `_empty_shapes_layer_selected` function gets called every time a layerlist selection event fires, and updates the value of the context key to a new instance of `fun` that is correctly parameterized with the current selection. Then, when the `ContextMapping` object gets created (in our case on right-click of layer), `fun` gets evaluated and we get the bool we were after.

I'm not sure if I love this hybrid approach. It's indirect and going to be confusing for anyone who comes across it later (regardless of how well we document it) - especially because for other functional keys, they are defined with `parse_expression` or explicitly assigned elsewhere, while the definition for this context key just looks "normal". Having said that, I wouldn't really say I have a better alternative? We could turn this into a function that takes no arguments and gets the layerlist from `provide_viewer`, but that would irreversibly make the layerlist require a fully built viewer for full functionality so I think it's almost a non-starter.

@lucyleeow has correctly pointed out that we actually could just **not change this context key at all** if we're concerned about the hybrid keys, since anyway there's other keys to fix. However, maybe it would be good to release a hybrid key (assuming it goes green with the `partial`) and seeing whether it creates any issues in the wild.

Reply:
To summarize, our options right now are to add the partial here, wait for CI to be green, and release **OR** make no changes whatsoever to `_selected_empty_shapes_layer` context key in this PR. I think...

Reply:
> they are defined with parse_expression

I think this is to avoid IDE errors see: https://github.com/napari/napari/pull/6864/files#r1602701616

Reply:
Yeah sorry, they could just be strings right? But I mean it's noticeably different from others where it's only ever a function

Reply:
@Czaki what do you think of @DragaDoncila's suggestion of using a partial instead of a nested function here? If you think it's ok, could you push here so we could merge?

@DragaDoncila I'm not sure what you mean by the distinction between "normal", "hybrid", and "functional" keys... It would be good to spell those out with some concrete examples showing the order of events given some example code/interactions...

Reply:
@jni yes, the different types of keys are a bit confusing, as is the distinction between them. I will open an issue to discuss I think, and expand a bit more with some examples. For now, if this goes green it can go in as far as I'm concerned!

Reply:
Partial will keep same reference as current implementation. But this leak is not caused by thi reference but because of some problem with PySide6:

```
  Traceback (most recent call last):
    File "/home/runner/work/napari/napari/napari/_qt/qt_main_window.py", line 335, in eventFilter
      return super().eventFilter(source, event)
  TypeError: 'PySide6.QtCore.QObject.eventFilter' called with wrong argument types:
    PySide6.QtCore.QObject.eventFilter(QWidgetItem, QEvent)
  Supported signatures:
    PySide6.QtCore.QObject.eventFilter(PySide6.QtCore.QObject, PySide6.QtCore.QEvent)
```

![obraz](https://github.com/napari/napari/assets/3826210/f6f2c6b7-25c6-4ba3-a0fb-991a8086c6a6)


[QtViewer-leak-backref-graph-1.pdf](https://github.com/user-attachments/files/16097416/QtViewer-leak-backref-graph-1.pdf)

Reply:
@Czaki I'm trying to interpret the graph — so what you're saying is, we/qtpy (?) passed the wrong type to QObject.eventFilter, a QWidgetItem (which is *not* a QObject for some reason???), *therefore* PySide2 raises, and *in the traceback*, there is a reference to the object, which we count as a leak? Should that be counted as a leak? 🤔

I think the version of the code with the partial is indeed more readable, but I'm happy for that to be a small follow-up PR.

Reply:
`PySide6` not `PySide2`. And it looks like QWidgetItem is not a subclass of QObject https://doc.qt.io/qt-6/qwidgetitem.html

It looks like qt main event loop pass event to QMainWindow.eventFilter, that we try to pass to default implementation, and then PySide6 wrapper fails to accept it. 

> Should that be counted as a leak? 🤔

No. It is also common pattern to report leak when there is an exception in the tests. The main problem is That I do not know how to filter out such cases. 



> I think the version of the code with the partial is indeed more readable, but I'm happy for that to be a small follow-up PR.

For me not. But if you prefer such a solution, we should use weakref 


```python

def _selection_check(s: LayerSel) -> bool:
	s_ = s()
	if s_ is None:
        return False
    return any(x._type_string == 'shapes' and not len(x.data) for x in s_)


def _empty_shapes_layer_selected(s: LayerSel) -> Callable[[], bool]:
    return partial(_selection_check, weakref.ref(s))

```    

Should I push this change?

Reply:
Out of interest, is there a risk of leak via this nested function/partial?

Reply:
Yeah I'm confused, because without fail, removing the nested function (whether it be with a pure function, or with a partial), fixed CI. But of course that could just be a fluke :woman_shrugging:

Reply:
> For me not. But if you prefer such a solution, we should use weakref

Is the typing incorrect, or should it be something like:

```python
def _selection_check(s: weakref[LayerSel]) -> bool:
```

?

other than that I like it, but I think I will leave it for a subsequent PR, maybe even for 0.5.1. Actually now might be a good time to introduce our own version of partial (`weak_partial`?) that wraps all the arguments in weakrefs.

btw, side question about weakrefs: I see we always check whether the weakref has been cleared by checking if the result is None — but what if the value being weakref'd is itself None? 🤔 In some cases, the desired behaviour might be different...

> It looks like qt main event loop pass event to QMainWindow.eventFilter, that we try to pass to default implementation, and then PySide6 wrapper fails to accept it.

This sounds like a bug in Qt/PySide6 but I'm not 100% sure from that description that it's not a bug in our use of eventFilter? Are you?

> No. It is also common pattern to report leak when there is an exception in the tests. The main problem is That I do not know how to filter out such cases.

I don't have time to look at the leak test but if we can have access to the graph we can maybe subtract the number of widgets accessible from sys, or the number of widgets accessible from sys *but not* from sys if you cut the traceback nodes out of the graph? I know it sounds complex but if we can come up with something reasonable and document it, it would probably save a lot of developer confusion in the future.

Anyway, I'm going to merge this, and we can update all the ref stuff in a subsequent PR. Thanks @Czaki!

Reply:
> This sounds like a bug in Qt/PySide6 but I'm not 100% sure from that description that it's not a bug in our use of eventFilter? Are you?

Yes. It is looking like not our bug, as we only pass this event to the super implementation. We do not convert it.

> btw, side question about weakrefs: I see we always check whether the weakref has been cleared by checking if the result is None — but what if the value being weakref'd is itself None? 🤔 In some cases, the desired behaviour might be different...

I think that it is not possible to make weakref to None. Especially, object needs to have `__weakref__` slot to allow for this. 

> Is the typing incorrect, or should it be something like:

I forget to update type annotation.